{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "11131a8b-5ee2-4d05-8e45-9e0affe1ffdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import socket\n",
    "import ssl\n",
    "from cryptography.x509 import load_pem_x509_certificate\n",
    "%run FunçõesAuxiliares.ipynb \n",
    "import os\n",
    "from cryptography import x509\n",
    "from cryptography.hazmat.backends import default_backend\n",
    "from cryptography.hazmat.primitives import hashes, serialization\n",
    "from cryptography.hazmat.primitives.asymmetric import rsa\n",
    "from cryptography.x509.oid import NameOID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af1b5af2-fd75-4ef6-b1a6-ec946d47f3c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== Cliente TLS com CSR e chat seguro ===\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Common Name (CN):  Lourenco\n",
      "Common Name (CN) do Destinatário:  Marianna\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Certificado já existe: certificados\\Lourenco_cert.pem\n"
     ]
    },
    {
     "ename": "TimeoutError",
     "evalue": "[WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTimeoutError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[4], line 136\u001b[0m\n\u001b[0;32m    133\u001b[0m     cert_dest \u001b[38;5;241m=\u001b[39m load_pem_x509_certificate(f\u001b[38;5;241m.\u001b[39mread())\n\u001b[0;32m    134\u001b[0m     chave_pub_dest \u001b[38;5;241m=\u001b[39m cert_dest\u001b[38;5;241m.\u001b[39mpublic_key()\n\u001b[1;32m--> 136\u001b[0m chat_continuo_cliente(\n\u001b[0;32m    137\u001b[0m     certfile\u001b[38;5;241m=\u001b[39mcaminho_cert_cliente,\n\u001b[0;32m    138\u001b[0m     keyfile\u001b[38;5;241m=\u001b[39mcaminho_key_cliente,\n\u001b[0;32m    139\u001b[0m     cafile\u001b[38;5;241m=\u001b[39mcaminho_cert_ca,\n\u001b[0;32m    140\u001b[0m     common_name_dest\u001b[38;5;241m=\u001b[39mcommon_name_dest,\n\u001b[0;32m    141\u001b[0m     caminho_cert_dest\u001b[38;5;241m=\u001b[39mcaminho_cert_dest,\n\u001b[0;32m    142\u001b[0m     chave_aes_servidor\u001b[38;5;241m=\u001b[39mchave_aes_servidor,\n\u001b[0;32m    143\u001b[0m     chave_pub_dest\u001b[38;5;241m=\u001b[39mchave_pub_dest\n\u001b[0;32m    144\u001b[0m )\n",
      "Cell \u001b[1;32mIn[4], line 73\u001b[0m, in \u001b[0;36mchat_continuo_cliente\u001b[1;34m(certfile, keyfile, cafile, common_name_dest, caminho_cert_dest, chave_aes_servidor, chave_pub_dest)\u001b[0m\n\u001b[0;32m     70\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mO Certificado Digital de quem deseja comunicar não passou as devidas validações. Conexão não pode ser iniciada.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m     71\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[1;32m---> 73\u001b[0m sock \u001b[38;5;241m=\u001b[39m socket\u001b[38;5;241m.\u001b[39mcreate_connection((\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m192.168.1.199\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;241m4443\u001b[39m))\n\u001b[0;32m     74\u001b[0m conn \u001b[38;5;241m=\u001b[39m context\u001b[38;5;241m.\u001b[39mwrap_socket(sock, server_hostname\u001b[38;5;241m=\u001b[39m common_name_dest)  \n\u001b[0;32m     76\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[1;32m~\\anaconda3\\Lib\\socket.py:865\u001b[0m, in \u001b[0;36mcreate_connection\u001b[1;34m(address, timeout, source_address, all_errors)\u001b[0m\n\u001b[0;32m    863\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m    864\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m all_errors:\n\u001b[1;32m--> 865\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m exceptions[\u001b[38;5;241m0\u001b[39m]\n\u001b[0;32m    866\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m ExceptionGroup(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcreate_connection failed\u001b[39m\u001b[38;5;124m\"\u001b[39m, exceptions)\n\u001b[0;32m    867\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[0;32m    868\u001b[0m     \u001b[38;5;66;03m# Break explicitly a reference cycle\u001b[39;00m\n",
      "File \u001b[1;32m~\\anaconda3\\Lib\\socket.py:850\u001b[0m, in \u001b[0;36mcreate_connection\u001b[1;34m(address, timeout, source_address, all_errors)\u001b[0m\n\u001b[0;32m    848\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m source_address:\n\u001b[0;32m    849\u001b[0m     sock\u001b[38;5;241m.\u001b[39mbind(source_address)\n\u001b[1;32m--> 850\u001b[0m sock\u001b[38;5;241m.\u001b[39mconnect(sa)\n\u001b[0;32m    851\u001b[0m \u001b[38;5;66;03m# Break explicitly a reference cycle\u001b[39;00m\n\u001b[0;32m    852\u001b[0m exceptions\u001b[38;5;241m.\u001b[39mclear()\n",
      "\u001b[1;31mTimeoutError\u001b[0m: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond"
     ]
    }
   ],
   "source": [
    "pasta_csrs = \"csrs\"\n",
    "pasta_certs = \"certificados\"\n",
    "pasta_ca = \"AutoridadeCertificadora\"\n",
    "os.makedirs(pasta_csrs, exist_ok=True)\n",
    "os.makedirs(pasta_certs, exist_ok=True)\n",
    "os.makedirs(pasta_ca, exist_ok=True)\n",
    "\n",
    "def gerar_chave_e_csr(common_name, country_name, state_or_province_name, locality_name, organization_name):\n",
    "    key_filename = os.path.join(pasta_certs, f\"{common_name}_chave.pem\")   # chave salva aqui\n",
    "    csr_filename = os.path.join(pasta_csrs, f\"{common_name}.csr\")    # CSR salva aqui\n",
    "\n",
    "    # Se já existe chave e csr, não gera de novo\n",
    "    if os.path.exists(key_filename) and os.path.exists(csr_filename):\n",
    "        print(f\"Chave e CSR para {common_name} já existem.\")\n",
    "        return key_filename, csr_filename\n",
    "\n",
    "    print(\"Gerando chave privada...\")\n",
    "    key = rsa.generate_private_key(\n",
    "        public_exponent=65537,\n",
    "        key_size=4096\n",
    "    )\n",
    "\n",
    "    print(\"Criando CSR...\")\n",
    "    subject = x509.Name([\n",
    "        x509.NameAttribute(NameOID.COUNTRY_NAME, country_name),\n",
    "        x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME, state_or_province_name),\n",
    "        x509.NameAttribute(NameOID.LOCALITY_NAME, locality_name),\n",
    "        x509.NameAttribute(NameOID.ORGANIZATION_NAME, organization_name),\n",
    "        x509.NameAttribute(NameOID.COMMON_NAME, common_name)\n",
    "    ])\n",
    "\n",
    "    csr = x509.CertificateSigningRequestBuilder().subject_name(subject).sign(key, hashes.SHA256(), default_backend())\n",
    "\n",
    "    with open(key_filename, \"wb\") as f:\n",
    "        f.write(key.private_bytes(\n",
    "            encoding=serialization.Encoding.PEM,\n",
    "            format=serialization.PrivateFormat.TraditionalOpenSSL,\n",
    "            encryption_algorithm=serialization.NoEncryption()\n",
    "        ))\n",
    "\n",
    "    with open(csr_filename, \"wb\") as f:\n",
    "        f.write(csr.public_bytes(serialization.Encoding.PEM))\n",
    "\n",
    "    print(f\"Chave e CSR criados: {key_filename}, {csr_filename}\")\n",
    "    print(f\"Envie o CSR (arquivo '{csr_filename}') para a CA para assinar o certificado.\")\n",
    "\n",
    "    return key_filename, csr_filename\n",
    "\n",
    "\n",
    "def esperar_certificado(common_name):\n",
    "    cert_filename = os.path.join(pasta_certs, f\"{common_name}_cert.pem\")\n",
    "    print(f\"\\nAguardando certificado assinado pela CA... ({cert_filename})\")\n",
    "\n",
    "    while not os.path.exists(cert_filename):\n",
    "        print(f\"Certificado ainda não encontrado. Por favor, copie o certificado emitido para este caminho: {cert_filename}\")\n",
    "        resposta = input(\"Pressione Enter para verificar novamente ou digite 'sair' para cancelar: \").strip().lower()\n",
    "        if resposta == 'sair':\n",
    "            print(\"Cancelando espera pelo certificado.\")\n",
    "            return None  # Retorna None para indicar cancelamento\n",
    "    print(f\"Certificado encontrado: {cert_filename}\")\n",
    "    return cert_filename\n",
    "\n",
    "def chat_continuo_cliente(certfile, keyfile, cafile, common_name_dest, caminho_cert_dest, chave_aes_servidor, chave_pub_dest):\n",
    "    context = ssl.create_default_context()\n",
    "    context.load_cert_chain(certfile=certfile, keyfile=keyfile)\n",
    "    context.load_verify_locations(cafile=cafile)\n",
    "    context.verify_mode = ssl.CERT_REQUIRED\n",
    "\n",
    "    if not(validacoes(caminho_cert_dest,cafile, common_name_dest)):\n",
    "        print('O Certificado Digital de quem deseja comunicar não passou as devidas validações. Conexão não pode ser iniciada.')\n",
    "        return\n",
    "\n",
    "    sock = socket.create_connection(('192.168.1.199', 4443))\n",
    "    conn = context.wrap_socket(sock, server_hostname= common_name_dest)  \n",
    "\n",
    "    try:\n",
    "        while True:\n",
    "            texto = input(\"Você: \")\n",
    "            if texto.lower() == \"sair\":\n",
    "                break\n",
    "\n",
    "            mensagem_criptografada = enviar_pacote_json(chave_aes_servidor, texto, chave_pub_dest)\n",
    "            print('\\n',mensagem_criptografada,'\\n')\n",
    "            conn.sendall(mensagem_criptografada)\n",
    "\n",
    "            resposta_criptografada = conn.recv(4096)\n",
    "            resposta = desencriptar_mensagem(resposta_criptografada.decode(), keyfile)\n",
    "            print(f\"Servidor diz: {resposta}\")\n",
    "\n",
    "    finally:\n",
    "        conn.close() \n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    print(\"=== Cliente TLS com CSR e chat seguro ===\")\n",
    "\n",
    "    common_name = input(\"Common Name (CN): \")\n",
    "    common_name_dest = input(\"Common Name (CN) do Destinatário: \")\n",
    "\n",
    "    caminho_cert_cliente = os.path.join(pasta_certs, f\"{common_name}_cert.pem\")\n",
    "    caminho_key_cliente = os.path.join(pasta_certs, f\"{common_name}_chave.pem\")\n",
    "    caminho_csr_cliente = os.path.join(pasta_csrs, f\"{common_name}.csr\")\n",
    "    chave_aes_servidor = gerar_chave_aes()\n",
    "    caminho_cert_ca = \"AutoridadeCertificadora/certificado_raiz.pem\"\n",
    "    caminho_cert_dest = os.path.join(pasta_certs, f\"{common_name_dest}_cert.pem\")\n",
    "\n",
    "    if not os.path.exists(caminho_cert_cliente):\n",
    "        # Se não tem certificado, então gera chave e CSR e espera pelo certificado\n",
    "        country_name = input(\"Country Name (2-letter code): \")\n",
    "        state_or_province_name = input(\"State or Province Name: \")\n",
    "        locality_name = input(\"Locality Name: \")\n",
    "        organization_name = input(\"Organization Name: \")\n",
    "\n",
    "        caminho_key_cliente, caminho_csr_cliente = gerar_chave_e_csr(\n",
    "            common_name, country_name, state_or_province_name, locality_name, organization_name\n",
    "        )\n",
    "        caminho_cert_cliente = esperar_certificado(common_name)\n",
    "        if caminho_cert_cliente is None:\n",
    "            print(\"Certificado não foi encontrado. Encerrando.\")\n",
    "            exit(0)\n",
    "    else:\n",
    "        print(f\"Certificado já existe: {caminho_cert_cliente}\")\n",
    "\n",
    "    if not os.path.exists(caminho_cert_ca):\n",
    "        print(f\"Erro: Certificado da CA raiz não encontrado em {caminho_cert_ca}.\")\n",
    "        exit(1)\n",
    "\n",
    "    if not os.path.exists(caminho_cert_dest):\n",
    "        print(f\"Erro: Certificado do destinatário não encontrado em {caminho_cert_dest}.\")\n",
    "        exit(1)\n",
    "\n",
    "    with open(caminho_cert_dest, \"rb\") as f:\n",
    "        cert_dest = load_pem_x509_certificate(f.read())\n",
    "        chave_pub_dest = cert_dest.public_key()\n",
    "\n",
    "    chat_continuo_cliente(\n",
    "        certfile=caminho_cert_cliente,\n",
    "        keyfile=caminho_key_cliente,\n",
    "        cafile=caminho_cert_ca,\n",
    "        common_name_dest=common_name_dest,\n",
    "        caminho_cert_dest=caminho_cert_dest,\n",
    "        chave_aes_servidor=chave_aes_servidor,\n",
    "        chave_pub_dest=chave_pub_dest\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29abebaa-25b5-4db8-9960-6adedd0dd9f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ficheiro ficheiros_teste/teste.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf16bb7a-996e-4aec-925a-a3edf2e4615f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
